### PR TITLE
MH-12607 (Harvard DCE), Demux Operation

### PR DESCRIPTION
**Pull request by lkiesow**
This is an updated version of pull request #14 which @phoebem updated [here](https://github.com/lkiesow/opencast/pull/1). During my review of #14, I noticed a few issues for which I provided some patches which are included here.

To make the review easier, this pull request contains only the new demux operation of #14.

### CODE REVIEW

**Comment by lkiesow** [comment]
- @gregorydlogan, I already included fixes based on your review from yesterday.
- @phoebem, can you check if this works for you too.

**Comment by phoebem** [comment]
Good documentation clarifications.

Thanks Lars and Greg.
Phoebe.

**Comment by staubesv** [comment]
As discussed with @lkiesow and @gregorydlogan, I'm taking over this review.

File: docs/guides/admin/docs/workflowoperationhandlers/demux-woh.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+-----------
+
+The demux operation can be used to demux multiple streams (e.g. presenter and presentation) from one container and put
+them into separate tracks. It uses a special encoding profile that has two outputs. It flavors the target media in the
+order listed in the encoding profile output.
+
+Parameter Table
+---------------
+
+|configuration keys | example                     | description                                                         |
```
Comment (staubesv) [inline docs/guides/admin/docs/workflowoperationhandlers/demux-woh.md]: "|" on next line?

Comment (staubesv) [inline docs/guides/admin/docs/workflowoperationhandlers/demux-woh.md]: I would prefer capital words in titles, but no strong opinion

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    Collection<Track> sourceTracks = elementSelector.select(mediaPackage, false);
+    if (sourceTracks.isEmpty()) {
+      logger.info("No matching tracks found");
+      return createResult(mediaPackage, Action.CONTINUE);
+    }
+
+    long totalTimeInQueue = 0;
+    Map<Job, Track> encodingJobs = new HashMap<>();
+    for (Track track : sourceTracks) {
+      logger.info("Encoding track {} using encoding profile '{}'", track, profile);
```
Comment (staubesv) [inline modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java]: Demuxing track...

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+        throw new WorkflowOperationException(String.format("Number of target flavors (%d) and output tracks (%d) do "
+                + "not match", targetFlavors.size(), composedTracks.size()));
+      }
+      if (composedTracks.size() != targetTags.length && targetTags.length != 1 && targetTags.length != 0) {
+        throw new WorkflowOperationException(String.format("Number of target tag groups (%d) and output tracks (%d) "
+                + "do not match", targetTags.length, composedTracks.size()));
+      }
+
+      // Flavor each track in the order read
+      int i = 0;
```
Comment (staubesv) [inline modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java]: Maybe flavorIndex and tagsIndex for sake of readability?

File: modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+      int j = 0;
+      for (Track composedTrack : composedTracks) {
+        // set flavor to the matching flavor in the order listed
+        composedTrack.setFlavor(newFlavor(sourceTrack, targetFlavors.get(i)));
+        if (targetFlavors.size() > 1) {
```
Comment (staubesv) [inline modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java]: Somewhat weird to read as based on the fact that at most to composedTracks are supported which is not obvious if you only look at this loop.

File: docs/guides/admin/docs/workflowoperationhandlers/demux-woh.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
+Demux Workflow Operation
+========================
+
+Description
+-----------
+
+The demux operation can be used to demux multiple streams (e.g. presenter and presentation) from one container and put
```
Comment (staubesv) [inline docs/guides/admin/docs/workflowoperationhandlers/demux-woh.md:7]: "multiple streams" or "at most two streams"?

**Review by staubesv — CHANGES_REQUESTED** [review]

**Comment by phoebem** [comment]
> On May 15, 2018, at 5:39 AM, Sven Stauber <notifications@github.com> wrote:
> 
> @staubesv requested changes on this pull request.
> 
> In docs/guides/admin/docs/workflowoperationhandlers/demux-woh.md <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_172-23discussion-5Fr188211177&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=stg-49KCTICuhrbJdYerQZHBQUounHE9X_J0rGzm5jw&s=6LFtuYrTIusMo7_CsMxuWYtrTk0hvOuyAcp9aOFQEKY&e=>:
> 
> > @@ -0,0 +1,54 @@
> +Demux Workflow Operation
> +========================
> +
> +Description
> +-----------
> +
> +The demux operation can be used to demux multiple streams (e.g. presenter and presentation) from one container and put
> +them into separate tracks. It uses a special encoding profile that has two outputs. It flavors the target media in the
> +order listed in the encoding profile output.
> +
> +Parameter Table
> +---------------
> +
> +|configuration keys | example                     | description                                                         |
> "|" on next line?
> 
> In docs/guides/admin/docs/workflowoperationhandlers/demux-woh.md <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_172-23discussion-5Fr188211613&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=stg-49KCTICuhrbJdYerQZHBQUounHE9X_J0rGzm5jw&s=AshTbbJCwnQgYV11QOKP99MDriYToPt-QMogZ6rtrfY&e=>:
> 
> > @@ -0,0 +1,54 @@
> +Demux Workflow Operation
> +========================
> +
> +Description
> +-----------
> +
> +The demux operation can be used to demux multiple streams (e.g. presenter and presentation) from one container and put
> +them into separate tracks. It uses a special encoding profile that has two outputs. It flavors the target media in the
> +order listed in the encoding profile output.
> +
> +Parameter Table
> +---------------
> +
> +|configuration keys | example                     | description                                                         |
> I would prefer capital words in titles, but no strong opinion
> 
> In modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_172-23discussion-5Fr188222369&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=stg-49KCTICuhrbJdYerQZHBQUounHE9X_J0rGzm5jw&s=fOvBjkhEuJGJf5OCoAkSAxG2ckXxedHO3voeS8EP5W8&e=>:
> 
> > +    // Find the encoding profile - should only be one
> +    EncodingProfile profile = composerService.getProfile(encodingProfile);
> +    if (profile == null) {
> +      throw new WorkflowOperationException(String.format("Encoding profile '%s' was not found", encodingProfile));
> +    }
> +    // Look for elements matching the tag
> +    Collection<Track> sourceTracks = elementSelector.select(mediaPackage, false);
> +    if (sourceTracks.isEmpty()) {
> +      logger.info("No matching tracks found");
> +      return createResult(mediaPackage, Action.CONTINUE);
> +    }
> +
> +    long totalTimeInQueue = 0;
> +    Map<Job, Track> encodingJobs = new HashMap<>();
> +    for (Track track : sourceTracks) {
> +      logger.info("Encoding track {} using encoding profile '{}'", track, profile);
> Demuxing track...
> 
> In modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_172-23discussion-5Fr188223379&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=stg-49KCTICuhrbJdYerQZHBQUounHE9X_J0rGzm5jw&s=MrVmAydUCPRa1S2rl0ERqUgMP0UYO7F8iPdKFEvAxlw&e=>:
> 
> > +        logger.warn("No output from Demux operation");
> +        continue;
> +      }
> +
> +      List<Track> composedTracks = (List<Track>) MediaPackageElementParser.getArrayFromXml(job.getPayload());
> +      if (composedTracks.size() != targetFlavors.size() && targetFlavors.size() != 1) {
> +        throw new WorkflowOperationException(String.format("Number of target flavors (%d) and output tracks (%d) do "
> +                + "not match", targetFlavors.size(), composedTracks.size()));
> +      }
> +      if (composedTracks.size() != targetTags.length && targetTags.length != 1 && targetTags.length != 0) {
> +        throw new WorkflowOperationException(String.format("Number of target tag groups (%d) and output tracks (%d) "
> +                + "do not match", targetTags.length, composedTracks.size()));
> +      }
> +
> +      // Flavor each track in the order read
> +      int i = 0;
> Maybe flavorIndex and tagsIndex for sake of readability?
> 
> In modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_172-23discussion-5Fr188223726&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=stg-49KCTICuhrbJdYerQZHBQUounHE9X_J0rGzm5jw&s=68rGGdSlNyl2KaZOwy0137_vuQZ5lOZjFIZkNR4XK-4&e=>:
> 
> > +      if (composedTracks.size() != targetFlavors.size() && targetFlavors.size() != 1) {
> +        throw new WorkflowOperationException(String.format("Number of target flavors (%d) and output tracks (%d) do "
> +                + "not match", targetFlavors.size(), composedTracks.size()));
> +      }
> +      if (composedTracks.size() != targetTags.length && targetTags.length != 1 && targetTags.length != 0) {
> +        throw new WorkflowOperationException(String.format("Number of target tag groups (%d) and output tracks (%d) "
> +                + "do not match", targetTags.length, composedTracks.size()));
> +      }
> +
> +      // Flavor each track in the order read
> +      int i = 0;
> +      int j = 0;
> +      for (Track composedTrack : composedTracks) {
> +        // set flavor to the matching flavor in the order listed
> +        composedTrack.setFlavor(newFlavor(sourceTrack, targetFlavors.get(i)));
> +        if (targetFlavors.size() > 1) {
> Somewhat weird to read as based on the fact that at most to composedTracks are supported which is not obvious if you only look at this loop.
> 
> In docs/guides/admin/docs/workflowoperationhandlers/demux-woh.md <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_opencast_opencast_pull_172-23discussion-5Fr188224691&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=TJQ-1SstdvXiL3rEjhdQ2hxEsqTStc7HrWDIrl0lxlI&m=stg-49KCTICuhrbJdYerQZHBQUounHE9X_J0rGzm5jw&s=muWKOdqaVcZJQ5c6vYGJhETu3AOy0ZW4zNV8gzJaoOE&e=>:
> 
> > @@ -0,0 +1,54 @@
> +Demux Workflow Operation
> +========================
> +
> +Description
> +-----------
> +
> +The demux operation can be used to demux multiple streams (e.g. presenter and presentation) from one container and put
> "multiple streams" or "at most two streams”?
> 
There is nothing to limit the number of streams if you have a workflow that can support it.
The number of target flavors/tags must match the number of target outputs in the encoding profile so that they can be identified and flavored.

**Comment by staubesv** [comment]
Hi @lkiesow 

Note that the merge of PR #171 has caused conflicts in this PR.

**Comment by lkiesow** [comment]
I'm trying to fix the conflicts and test failures and resolve your comments until tomorrow. Will let you know once I'm done.

**Comment by lkiesow** [comment]
Done.

**Review by staubesv — APPROVED** [review]

**Pull request merged**